src/compiler/extensions.ts

```diff
@@ -0,0 +1,379 @@
+         * @param {string} err The error message to report
+         * @param {Node} span The node on which to position the error
+         */
+        (level: DiagnosticCategory, err: string, span: Node): void;
+        /**
+         * @param {DiagnosticCategory} level The error level to report this error as (Message, Warning, or Error)
+         * @param {string} err The error message to report
+         * @param {number} start The start position of the error span
+         * @param {number} length The length of the error span
+         */
+        (level: DiagnosticCategory, err: string, start: number, length: number): void;
+        /**
+         * @param {DiagnosticCategory} level The error level to report this error as (Message, Warning, or Error)
+         * @param {string} shortname A short code uniquely identifying the error within the lint
+         * @param {string} err The error message to report
+         */
+        (level: DiagnosticCategory, shortname: string, err: string): void;
+        /**
+         * @param {DiagnosticCategory} level The error level to report this error as (Message, Warning, or Error)
+         * @param {string} shortname A short code uniquely identifying the error within the lint
+         * @param {string} err The error message to report
+         * @param {Node} span The node on which to position the error
+         */
+        (level: DiagnosticCategory, shortname: string, err: string, span: Node): void;
+        /**
+         * @param {DiagnosticCategory} level The error level to report this error as (Message, Warning, or Error)
+         * @param {string} shortname A short code uniquely identifying the error within the lint
+         * @param {string} err The error message to report
+         * @param {number} start The start position of the error span
+         * @param {number} length The length of the error span
+         */
+        (level: DiagnosticCategory, shortname: string, err: string, start: number, length: number): void;
+    };
+
+    /*
+    * Walkers call stop to halt recursion into the node's children
+    * Walkers call error to add errors to the output.
+    */
+    export interface LintWalker {
+        /**
+         * @param {Node} node The current node being visited (starts at every SourceFile and recurs into their children)
+         * @param {LintErrorMethod} error A callback to add errors to the output
+         * @returns boolean true if this lint no longer needs to recur into the active node
+         */
+        visit(node: Node, error: LintErrorMethod): boolean | void;
+        /**
+         * Yar
+         * @param {Node} node The current node which has just finished being visited
+         * @param {LintErrorMethod} error A callback to add errors to the output
+         */
+        afterVisit?(node: Node, error: LintErrorMethod): void;
+    }
+
+    export interface BaseProviderStatic {
+        readonly ["extension-kind"]: ExtensionKind;

```

Comment:
Honestly? Because I think `extensionKind` is comparably ugly when used to annotate functions. Look at this:
```ts
function LintWalk() {
  // ...
}
LintWalk.extensionKind = "semantic-lint"
```
vs
```ts
function LintWalk() {
  // ...
}
LintWalk["extension-kind"] = "semantic-lint"
```
 Almost anything else I say would just be ex-facto justification for that. Purely stylistic choice. There's only one name we force extension API consumers to use, and it's this. Placing its canonical usage into a string literal easily exempts it from their own style checks, which can make it easier to handle when you have to implement something where its style doesn't match yours.

I think I gave another reason elsewhere in this thread, too, but I can't find it.